examples/computer_vision/vision_transformer/det_callback.py

```diff
@@ -0,0 +1,150 @@
+import typing
+
+from transformers import (
+    TrainerCallback,
+    TrainerState,
+    TrainerControl,
+    TrainingArguments,
+)
+
+import os
+import determined as det
+import logging
+
+import weakref
+import importlib.util
+
+
+class DetCallback(TrainerCallback):
+
+    def __init__(self, args: typing.Dict, filter_metrics: typing.List = None,
+                 tokenizer: typing.Any = None, tokenizer_options: typing.Dict = None,
+                 checkpoint_metadata: typing.Dict = None) -> None:
+        super().__init__()
+
+        assert is_determined_available(), "DetCallback requires determined to be installed. Run `pip install determined`."
+        import determined
+        self._det = determined
+        distributed = self._det.core.DistributedContext.from_torch_distributed()
+        self.core_context = self._det.core.init(distributed=distributed)
+        self.core_context.__enter__()
+        weakref.finalize(self, exit_context, self.core_context)
+
+        self.filter_metrics = filter_metrics
+        self.user_checkpoint_metadata = checkpoint_metadata
+        self.tokenizer = tokenizer
+        self.tokenizer_options = tokenizer_options
+        self.load_last_checkpoint(args)
+
+        self.last_eval_metrics = None
+        self.searcher_metric = self._det.get_cluster_info().trial._config['searcher']['metric']
+        self.searcher_ops = self.core_context.searcher.operations()
+        self.current_op = next(self.searcher_ops)
+
+    def on_log(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, model=None, logs=None,
+               **kwargs):
+        if state.is_world_process_zero:
+            metrics = self.process_log(logs)
+            if 'loss' in metrics:
+                self.core_context.train.report_training_metrics(steps_completed=state.global_step, metrics=metrics)
+            elif 'eval_loss' in metrics:
+                self.core_context.train.report_validation_metrics(steps_completed=state.global_step, metrics=metrics)
+                self.last_eval_metrics = metrics
+            else:
+                logging.warning(
+                    f"Metrics not reported: unknown metric type in {logs}.")
+
+    def process_log(self, logs: typing.Dict) -> typing.Dict:
+        metrics = logs
+
+        if self.filter_metrics:
+            metrics = {}
+            for k, v in logs.items():
+                if any(m in k for m in self.filter_metrics) is True:
+                    metrics[k] = v
+
+        return metrics
+
+    def on_save(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+        info = self._det.get_cluster_info()
+        assert info is not None
```

Comment:
The `info is None` case should probably skip saving checkpoints to the store instead of erroring, to support local mode in the future.